src/storage/storage.cpp

```diff
@@ -184,643 +182,670 @@ Storage::ReturnCode Storage::Run(int max_wait)
-    shared_layout_ptr->SetBlockSize<util::guidance::LaneTupleIdPair>(
-        SharedDataLayout::TURN_LANE_DATA, lane_tupel_count);
+    // Load the HSGR file
+    {
+        io::File hsgr_file(config.hsgr_data_path);
+        auto hsgr_header = io::readHSGRHeader(hsgr_file);
+        unsigned *checksum_ptr =
+            layout_ptr->GetBlockPtr<unsigned, true>(memory_ptr, DataLayout::HSGR_CHECKSUM);
+        *checksum_ptr = hsgr_header.checksum;
 
-    if (!static_cast<bool>(intersection_stream))
-        throw util::exception("Failed to read bearing values from " +
-                              config.intersection_class_path.string());
+        // load the nodes of the search graph
+        QueryGraph::NodeArrayEntry *graph_node_list_ptr =
+            layout_ptr->GetBlockPtr<QueryGraph::NodeArrayEntry, true>(memory_ptr,
+                                                                      DataLayout::GRAPH_NODE_LIST);
 
-    std::vector<util::guidance::EntryClass> entry_class_table;
-    if (!util::deserializeVector(intersection_stream, entry_class_table))
-        throw util::exception("Failed to read entry classes from " +
-                              config.intersection_class_path.string());
+        // load the edges of the search graph
+        QueryGraph::EdgeArrayEntry *graph_edge_list_ptr =
+            layout_ptr->GetBlockPtr<QueryGraph::EdgeArrayEntry, true>(memory_ptr,
+                                                                      DataLayout::GRAPH_EDGE_LIST);
 
-    shared_layout_ptr->SetBlockSize<util::guidance::EntryClass>(SharedDataLayout::ENTRY_CLASS,
-                                                                entry_class_table.size());
+        io::readHSGR(hsgr_file,
+                     graph_node_list_ptr,
+                     hsgr_header.number_of_nodes,
+                     graph_edge_list_ptr,
+                     hsgr_header.number_of_edges);
+    }
 
-    // allocate shared memory block
-    util::SimpleLogger().Write() << "allocating shared memory of "
-                                 << shared_layout_ptr->GetSizeOfLayout() << " bytes";
-    auto shared_memory = makeSharedMemory(data_region, shared_layout_ptr->GetSizeOfLayout(), true);
-    char *shared_memory_ptr = static_cast<char *>(shared_memory->Ptr());
+    // store the filename of the on-disk portion of the RTree
+    {
+        const auto file_index_path_ptr =
+            layout_ptr->GetBlockPtr<char, true>(memory_ptr, DataLayout::FILE_INDEX_PATH);
+        // make sure we have 0 ending
+        std::fill(file_index_path_ptr,
+                  file_index_path_ptr + layout_ptr->GetBlockSize(DataLayout::FILE_INDEX_PATH),
+                  0);
+        auto absolute_file_index_path = boost::filesystem::absolute(config.file_index_path);
+        std::copy(absolute_file_index_path.string().begin(),
+                  absolute_file_index_path.string().end(),
```

Comment:
This is Undefined Behavior  - string() returns a temporary so .begin() and .end() point to different variables.